File: src/programmers/skillcheck/level3/Palindrome.java
package programmers.skillcheck.level3;

import java.util.Optional;

/**
 * 프로그래머스
 * 스킬체크 레벨 3
 *
 * Palindrome
 * chanllenge_id=318
 *
 * 문자열 s에서, s의 부분문자열중
 * 가장 긴 팰린드롬의 길이를 return 하는 함수 만들기
 */
public class Palindrome {

    public int solution(String s) {

        // s의 substring의 길이를 줄여가면서 검사
        for (int i = s.length(); i > 0; i--) {

            int repeat = 1;
            // s의 substring중 길이가 i인 것을 검사
            for (int j = 0; j < repeat; j++) {
                String substring = s.substring(j, i + j);
                if (checkPalindrome(substring)) {
                    return substring.length();
                }
            }
            repeat++;
        }
        return 0;
    }

    public boolean checkPalindrome(String s) {
        Optional<String> maybeString = Optional.ofNullable(s);

        if (maybeString.isPresent()) return false;

        String reversed = new StringBuffer(maybeString.get()).reverse().toString();
        return maybeString.get().equals(reversed);
    }

    public static void main(String[] args) {
        Palindrome obj = new Palindrome();

        String s_1 = "abcdcba";
        int answer_1 = obj.solution(s_1);
        System.out.println("########## 예제 1의 결과는 ?");
        System.out.println(answer_1);
        System.out.println("##########");

        String s_2 = "abacde";
        int answer_2 = obj.solution(s_2);
        System.out.println("########## 예제 2의 결과는 ?");
        System.out.println(answer_2);
        System.out.println("##########");

    }
}


/*
채점을 시작합니다.
정확성  테스트
테스트 1 〉	실패 (1.81ms, 52.3MB)
테스트 2 〉	실패 (1.86ms, 50MB)
테스트 3 〉	실패 (2.47ms, 53MB)
테스트 4 〉	실패 (2.27ms, 50.1MB)
테스트 5 〉	실패 (2.39ms, 52.3MB)
테스트 6 〉	실패 (2.35ms, 50.5MB)
테스트 7 〉	실패 (2.41ms, 50.3MB)
테스트 8 〉	실패 (2.26ms, 54.1MB)
테스트 9 〉	실패 (1.91ms, 50.6MB)
테스트 10 〉	실패 (1.74ms, 52.2MB)
테스트 11 〉	통과 (0.98ms, 50.3MB)
테스트 12 〉	통과 (0.98ms, 50.6MB)
테스트 13 〉	실패 (2.38ms, 52.8MB)
테스트 14 〉	실패 (2.67ms, 53.9MB)
테스트 15 〉	실패 (2.49ms, 50.7MB)
테스트 16 〉	실패 (2.78ms, 52.2MB)
테스트 17 〉	통과 (0.64ms, 50.1MB)
테스트 18 〉	통과 (1.11ms, 54.6MB)
테스트 19 〉	실패 (2.22ms, 52.5MB)
테스트 20 〉	통과 (1.05ms, 52.5MB)
테스트 21 〉	실패 (2.06ms, 52.2MB)
효율성  테스트
테스트 1 〉	실패 (19.89ms, 58.7MB)
테스트 2 〉	통과 (1.06ms, 49.7MB)
채점 결과
정확성: 8.3
효율성: 7.7
합계: 15.9 / 50
 */

File: src/programmers/skillcheck/level3/Palindrome.js
/**
 * 프로그래머스
 * 스킬체크 레벨 3
 *
 * 문자열 s에서, s의 부분문자열중
 * 가장 긴 팰린드롬의 길이를 return 하는 함수 만들기
 * @param {String} s
 */
function solution(s) {
  // 순차 탐색으로 찾아내기?
  let start = 0;
  let end = 1;

  const candidates = [];

  while (true) {
    const test = s.substring(start, end);
    if (checkPalindrome(test)) {
      candidates.push(test.length);
    }

    // end가 끝에 도달
    end++;
    if (end > s.length) {
      start++;
      end = start + 1;
      if (start > s.length) {
        break;
      }
    }
  }

  console.log(candidates);
  return Math.max(...candidates);
}
function checkPalindrome(str) {
  return str == str.split("").reverse().join("");
}

/**
 * 2020 - 01 - 18
 * O(n^2) 이고 n <= 2500 이라 효율성을 통과 못한듯하다
 * 투 포인터로 될까 고민해보고 있다..
 * 
 * 채점을 시작합니다.
정확성  테스트
테스트 1 〉	통과 (0.18ms, 30.1MB)
테스트 2 〉	통과 (0.18ms, 30.1MB)
테스트 3 〉	통과 (6.74ms, 31.6MB)
테스트 4 〉	통과 (6.61ms, 31.9MB)
테스트 5 〉	통과 (6.81ms, 31.6MB)
테스트 6 〉	통과 (6.68ms, 31.6MB)
테스트 7 〉	통과 (6.29ms, 31.8MB)
테스트 8 〉	통과 (6.00ms, 32MB)
테스트 9 〉	통과 (6.13ms, 31.8MB)
테스트 10 〉	통과 (5.92ms, 31.4MB)
테스트 11 〉	통과 (6.83ms, 33.1MB)
테스트 12 〉	통과 (8.00ms, 33.8MB)
테스트 13 〉	통과 (5.99ms, 31.2MB)
테스트 14 〉	통과 (34.41ms, 34.6MB)
테스트 15 〉	통과 (30.73ms, 35.1MB) 
테스트 16 〉	통과 (30.37ms, 34.5MB)
테스트 17 〉	통과 (0.10ms, 30MB)
테스트 18 〉	통과 (0.11ms, 30.1MB)
테스트 19 〉	통과 (5.58ms, 31.5MB)
테스트 20 〉	통과 (6.75ms, 31.6MB)
테스트 21 〉	통과 (6.77ms, 31.7MB)
효율성  테스트
테스트 1 〉	실패 (시간 초과)
테스트 2 〉	실패 (시간 초과)
채점 결과
정확성: 34.7
효율성: 0.0
합계: 34.7 / 50
 */
